### PR TITLE
[Hail][feature] add `outer` option to `union_cols`

### LINKED ISSUE #7465: Add ability to outer join on rows in `union_cols`

### PR DESCRIPTION
**Pull request by patrick-schultz**
closes #7465

### CODE REVIEW

File: hail/python/hail/matrixtable.py [hunk 1]
```diff
@@ -3570,8 +3570,8 @@ def union_rows(*datasets: 'MatrixTable', _check_cols=True) -> 'MatrixTable':
                                      f"Datasets 0 and {wrong_keys+1} have different columns (or possibly different order).")
             return MatrixTable(MatrixUnionRows(*[d._mir for d in datasets]))
 
-    @typecheck_method(other=matrix_table_type)
-    def union_cols(self, other: 'MatrixTable') -> 'MatrixTable':
+    @typecheck_method(other=matrix_table_type, _outer=bool)
+    def union_cols(self, other: 'MatrixTable', _outer=False) -> 'MatrixTable':
```
Comment (tpoterba) [inline hail/python/hail/matrixtable.py]: why not make this public?

File: hail/python/test/hail/matrixtable/test_matrix_table.py [hunk 2]
```diff
@@ -530,6 +530,11 @@ def test_union_cols_distinct(self):
         mt = mt.key_rows_by(x = mt.row_idx // 2)
         assert mt.union_cols(mt).count_rows() == 5
 
+    def test_union_cols_outer(self):
```
Comment (tpoterba) [inline hail/python/test/hail/matrixtable/test_matrix_table.py]: can you add a test for correct entry joining?

File: hail/python/hail/matrixtable.py (hunk 1 above)
Comment (tpoterba) [reply hail/python/hail/matrixtable.py]: Just need a better test and some docs and we can make it public.

Comment (patrick-schultz) [reply hail/python/hail/matrixtable.py]: Sure. I was just being cautious until we'd had more discussion about the interface. Do you like `outer`, or should it be more explicit like `outer_join_rows`?

Comment (tpoterba) [reply hail/python/hail/matrixtable.py]: I like `row_join_type` or something that takes a string, with a default of inner.

That way we can add left/right if people want that (reasonable)

Comment (patrick-schultz) [reply hail/python/hail/matrixtable.py]: Seems good to me

File: hail/python/hail/matrixtable.py [hunk 3]
```diff
@@ -3570,8 +3570,9 @@ def union_rows(*datasets: 'MatrixTable', _check_cols=True) -> 'MatrixTable':
                                      f"Datasets 0 and {wrong_keys+1} have different columns (or possibly different order).")
             return MatrixTable(MatrixUnionRows(*[d._mir for d in datasets]))
 
-    @typecheck_method(other=matrix_table_type)
-    def union_cols(self, other: 'MatrixTable') -> 'MatrixTable':
+    @typecheck_method(other=matrix_table_type,
+                      row_join_type=enumeration('inner', 'outer', 'left', 'right'))
```
Comment (tpoterba) [inline hail/python/hail/matrixtable.py]: you don't support left/right

File: hail/python/hail/matrixtable.py [hunk 4]
```diff
@@ -3593,10 +3594,22 @@ def union_cols(self, other: 'MatrixTable') -> 'MatrixTable':
...
 
-        This method performs an inner join on rows and concatenates entries
-        from the two datasets for each row. Only distinct keys from each
-        dataset are included (equivalent to calling :meth:`.distinct_by_row`
-        on each dataset first).
+        This method creates a :class:`.MatrixTable` which contains all columns
+        from both input datasets. The set of rows included in the result is
+        determined by the `row_join_type` parameter.
+
+        - With the default ``row_join_type=inner``, an inner join is performed
```
Comment (tpoterba) [inline hail/python/hail/matrixtable.py]: We generally don't format the arg bit as code: should be something like:

    with the default value of ``'inner'``

File: hail/python/hail/matrixtable.py (hunk 3 above)
Comment (patrick-schultz) [reply hail/python/hail/matrixtable.py]: Eek, lazy copy/pasting

Comment (tpoterba) [reply hail/python/hail/matrixtable.py]: that's what review is for!

**Pull request merged**